test: interface_ipc.py minor fixes and cleanup

ryanofsky opened the pull request:
This PR cleans up the `interface_ipc.py` test, fixing broken checks, fixing missing await calls, removing to_dict calls, renaming variables, reducing `.result` accesses, and giving template objects explicit lifetimes. More details are in the commit messages.

The first commit changes a lot of indentation so is easiest to review ignoring whitespace.

ryanofsky committed:
test: fix interface_ipc.py template destruction

Use context managers to destroy block templates. Previously, block templates
were not being destroyed before disconnecting because the destroy coroutines
were called but never awaited. It's not necessary to explicitly destroy the
templates since they will get garbage collected asynchronously, but it's good
to destroy them to make the test more predictable, and to make the destroy
calls that are present actually do something.

This change also removes `await waitnext` expressions without changing
behavior, because the previous code was misleading about what order waitNext
calls were executed.

This change is easiest to review ignoring whitespace.

Co-authored-by: Sjors Provoost <sjors@sprovoost.nl>

sedited reviewed (approved):
ACK 6ef092c0e4343fc421c323ff09d3c791fb4bc33a

Thanks for the cleanup! Have another suggestion for saving a few lines and naming clarity:

<details>
<summary>Diff</summary>

```diff
diff --git a/test/functional/interface_ipc.py b/test/functional/interface_ipc.py
index 804f7f8f64..f8f3a76ec7 100755
--- a/test/functional/interface_ipc.py
+++ b/test/functional/interface_ipc.py
@@ -176,2 +176 @@ class IPCInterfaceTest(BitcoinTestFramework):
-                template3 = await template2.waitNext(ctx, waitoptions)
-                assert_equal(template3._has("result"), False)
+                assert_equal((await template2.waitNext(ctx, waitoptions))._has("result"), False)
@@ -181,2 +180,2 @@ class IPCInterfaceTest(BitcoinTestFramework):
-                template4 = await stack.enter_async_context(destroying((await template2.waitNext(ctx, waitoptions)).result, ctx))
-                block3 = await self.parse_and_deserialize_block(template4, ctx)
+                template3 = await stack.enter_async_context(destroying((await template2.waitNext(ctx, waitoptions)).result, ctx))
+                block3 = await self.parse_and_deserialize_block(template3, ctx)
@@ -187,2 +186,2 @@ class IPCInterfaceTest(BitcoinTestFramework):
-                template5 = await stack.enter_async_context(destroying((await template4.waitNext(ctx, waitoptions)).result, ctx))
-                block4 = await self.parse_and_deserialize_block(template5, ctx)
+                template4 = await stack.enter_async_context(destroying((await template3.waitNext(ctx, waitoptions)).result, ctx))
+                block4 = await self.parse_and_deserialize_block(template4, ctx)
@@ -194,3 +193,3 @@ class IPCInterfaceTest(BitcoinTestFramework):
-                template6 = await stack.enter_async_context(destroying((await template5.waitNext(ctx, waitoptions)).result, ctx))
-                block4 = await self.parse_and_deserialize_block(template6, ctx)
-                assert_equal(len(block4.vtx), 3)
+                template5 = await stack.enter_async_context(destroying((await template4.waitNext(ctx, waitoptions)).result, ctx))
+                block5 = await self.parse_and_deserialize_block(template5, ctx)
+                assert_equal(len(block5.vtx), 3)
@@ -199,2 +198 @@ class IPCInterfaceTest(BitcoinTestFramework):
-                template7 = await template6.waitNext(ctx, waitoptions)
-                assert_equal(template7._has("result"), False)
+                assert_equal((await template5.waitNext(ctx, waitoptions))._has("result"), False)
@@ -209,2 +207 @@ class IPCInterfaceTest(BitcoinTestFramework):
-                    template7 = await template6.waitNext(ctx, new_waitoptions)
-                    assert_equal(template7._has("result"), False)
+                    assert_equal((await template5.waitNext(ctx, new_waitoptions))._has("result"), False)
@@ -215 +212 @@ class IPCInterfaceTest(BitcoinTestFramework):
-                    template6.interruptWait()
+                    template5.interruptWait()

```

</details>

rkrux reviewed (commented):
Concept ACK 6ef092c0e4343fc421c323ff09d3c791fb4bc33a

I'm new to this test, I built the `bitcoin-node` executable using `-DENABLE_IPC=ON` and ran the test using `venv` as mentioned in the readme file. I spent some time trying to understand the interfaces involving capnp. 

The changes seem to clean the test quite a bit and makes it easier to go through.

mercie-ux commented:
Concept ACK  #34003  

I ran the interface_ipc.py test locally and it passes. The PR cleans up the test by fixing broken checks, missing await calls, and variable naming. 

I’m new to PR reviews, but from what I can tell, the changes seem correct and improve readability. I don’t see any issues with the changes.

ryanofsky committed:
test: interface_ipc.py minor fixes and cleanup

There are a few things that are incorrect or messy in the interface_ipc.py test.
This commit tries to clean them up:

- isTestChain and isInitialBlockDownload asserts were not checking the results
  of those calls, only that calls were, made because they were not checking the
  responses' .result member.

- A lot of result accesses like `template.result` `mining.result` were repeated
  unnecessarily because variables like `template` and `mining` were assigned
  response objects instead of result objects. These variables are now changed
  to point directly to results.

- Some coroutine calls were assigned to temporary `wait` before being awaited.
  This was unnecessarily confusing and would make code not run in top-down
  order.

- `to_dict` calls were being made to check if result variables were unset. This
  was inefficient and indirect because it iterates over all fields in response
  structs instead of just checking whether the result field is present. The
  to_dict calls are now replaced with more direct `_has('result')` calls.

- The `res` variables used to hold various responses did not have descriptive
  names. These are replaced with clearer names.

Co-authored-by: rkrux <rkrux.connect@gmail.com>

ryanofsky committed:
test: improve interface_ipc.py waitNext tests

As pointed out by Sjors in
https://github.com/bitcoin/bitcoin/pull/34003#discussion_r2598844209 and
https://github.com/bitcoin/bitcoin/pull/34003#discussion_r2598858386 the
original intention of having waitNext and waitTipChanged calls in the test was
to ensure that if new blocks were connected or fees were increased *during* the
waits, that the calls would wake up and return.

But the tests were written incorrectly, to generate blocks and transactions
before the wait calls instead of during the calls. So the tests were less
meaningful then they should be.

There was also a similar problem in the interruptWait test. The test was
intended to test the interruptWait method, but it was never actually calling
the method due to a missing await keyword. Instead it was testing that
miniwallet.send_self_transfer would interrupt the wait.

This commit fixes these issues by introducing a wait_and_do() helper function
to start parallel tasks and trigger an action after a wait call is started.

Co-authored-by: Sjors Provoost <sjors@sprovoost.nl>

ryanofsky commented on test/functional/interface_ipc.py:
re: https://github.com/bitcoin/bitcoin/pull/34003#discussion_r2592834249

Thanks, fixed message. I forgot to update it when the commit was split up.

ryanofsky commented on test/functional/interface_ipc.py:
```diff
@@ -134,126 +142,124 @@ async def async_routine():
+                waitoptions.timeout = timeout
+                waitoptions.feeThreshold = 1
+                self.generate(self.nodes[0], 1)
+                template2 = await stack.enter_async_context(destroying((await template.waitNext(ctx, waitoptions)).result, ctx))
```
re: https://github.com/bitcoin/bitcoin/pull/34003#discussion_r2598844209

Added a new commit to make this check work as originally intended and call generate during the wait instead of before.

ryanofsky reviewed (commented):
Thanks for the reviews!

<!-- begin push-3 -->
Updated 6ef092c0e4343fc421c323ff09d3c791fb4bc33a -> c358910b8cd6d1db51c10becb0cbcb58709b738f ([`pr/pyipc.2`](https://github.com/ryanofsky/bitcoin/commits/pr/pyipc.2) -> [`pr/pyipc.3`](https://github.com/ryanofsky/bitcoin/commits/pr/pyipc.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/pyipc.2..pr/pyipc.3))<!-- end --> applying suggestions and adding a new commit to test actions during waitNext calls instead of before them.

I tried to do the same thing with the waitTipChanged call as well, but for some reason waitTipChanged always seems to return right away even when the tip has not changed, and I'm not sure why that is, so I left it alone for now. Change I tried was:

```diff
-            self.generate(self.nodes[0], 1)
-            newblockref = (await mining.waitTipChanged(ctx, blockref.result.hash)).result
+            newblockref = (await wait_and_do(mining.waitTipChanged(ctx, blockref.result.hash), lambda: self.generate(self.nodes[0], 1))).result
```

sedited reviewed (approved):
ACK d8fe5f0326c5c0505e0f7e6e978961d24c27b175

rkrux commented on test/functional/interface_ipc.py:
```diff
@@ -38,6 +39,29 @@ async def wait_next_template(template, stack, ctx, opts):
+        else:
+            do_fn()
+
+    await asyncio.gather(wait(), do())
```
In d8fe5f0326c5c0505e0f7e6e978961d24c27b175 "test: improve interface_ipc.py waitNext tests"

All are nits now I believe.

1. I don't suppose the following is guaranteed? When event is set, then the wait of `do` is also over that makes it call the sleep in parallel. Reworded slightly to highlight the broader intent of the `wait_and_do` function.
> Call wait_fn, then sleep,

2. From an event's POV, it is set in `wait()` and waited for in `do()`; suggested `s/wait/setter` and `s/do/waiter`.

3. `s/wait_started/event_over` because I got confused by `wait_started.set()` that could also suggest that the event wait starts now when instead it gets over at this stage.

```diff
diff --git a/test/functional/interface_ipc.py b/test/functional/interface_ipc.py
index 7c8f51a81c..0da75d25cd 100755
--- a/test/functional/interface_ipc.py
+++ b/test/functional/interface_ipc.py
@@ -40,18 +40,18 @@ async def wait_next_template(template, stack, ctx, opts):
     return await stack.enter_async_context(destroying((await template.waitNext(ctx, opts)).result, ctx))
 
 async def wait_and_do(wait_fn, do_fn):
-    """Call wait_fn, then sleep, then call do_fn in a parallel task. Wait for
-    both tasks to complete."""
-    wait_started = asyncio.Event()
+    """Call wait_fn and do_fn in parallel tasks while ensuring do_fn is
+    called after wait_fn is started. Wait for both tasks to complete."""
+    event_over = asyncio.Event()
     result = None
 
-    async def wait():
+    async def setter():
         nonlocal result
-        wait_started.set()
+        event_over.set()
         result = await wait_fn
 
-    async def do():
-        await wait_started.wait()
+    async def waiter():
+        await event_over.wait()
         await asyncio.sleep(0.1)
         # Let do_fn be either a callable or an awaitable object
         if inspect.isawaitable(do_fn):
@@ -59,7 +59,7 @@ async def wait_and_do(wait_fn, do_fn):
         else:
             do_fn()
 
-    await asyncio.gather(wait(), do())
+    await asyncio.gather(setter(), waiter())
     return result
 
 class IPCInterfaceTest(BitcoinTestFramework):
```

rkrux reviewed (commented):
Post-merge now as it got merged while I was reviewing.

ACK d8fe5f0326c5c0505e0f7e6e978961d24c27b175

I have focussed only on the changes in this PR and have limited understanding of the mining/ipc domain.
